### PR TITLE
Topic/document save

### LINKED ISSUE #4597: sclang: add instance method to Document to save a file
##  Motivation

`Document` currently has many features to control the IDE, such as `Document.new` to open a new document in the IDE. Sometimes, the IDE crashes (not often!!!) and so saving the file could be implemented at the top of one's code to ensure changes aren't lost (maybe?). Personally, I have implemented an automatic git commit upon code evaluation, but this only works if the file has been saved.

## Description of Proposed Feature

To automatically have sclang save the current document, using something like: `Document.current.save`

## Plan for Implementation

This shouldn't take a ton of time. I have limited C++ experience, otherwise, I could do this myself (maybe I will when I get more time!), but maybe someone who knows C++ really well could implement this relatively quickly. I see that in doc_manager.cpp there is a function `handleScLangMessage` but "save" or "saveDocument" is not one of the sclang messages it responds to.

### PR DESCRIPTION
**Pull request by muellmusik**
## Purpose and Motivation

Fixes #4597. Adds a save method to the Document class in the IDE.

## Types of changes

- Documentation
- New feature

I didn't add a unit test for this as it seemed overkill and would mostly test dependant code, but can do so if that's preferred.

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by muellmusik** [comment]
Hmm Travis failed on OSX with:
```
Fetching submodule external_libraries/nova-tt
5754fatal: remote error: upload-pack: not our ref 4e307364528bd36b80fa61ccf2d3b60916dbba51
5755The command "git fetch --tags" failed and exited with 1 during .
```

and on linux with Linting on
```
The command "$TRAVIS_BUILD_DIR/.travis/before-install-$TRAVIS_OS_NAME.sh --qt=$QT" failed and exited with 1 during .
```

Advice gratefully received!

**Comment by muellmusik** [comment]
Hmm looks like similar issues in other PRs right now.

**Comment by jamshark70** [comment]
As much as I loathe writing unit tests, I do think it would be good to do a simple one here:

1. Create a document instance with a known string for contents.
2. `.save` it to a temp location.
3. Read it back with `File` and check the contents.
4. (Delete temp file.)

If something ever drifted out of whack in IDE save code, this might actually catch it (where currently we don't have any test for it).

**Comment by jamshark70** [comment]
Edited my last comment.

A unit test for this raises a sync issue: `Document.new` synchronously adds to the allDocuments collection -- so there's no way here to avoid a hardcoded wait as far as I can see. Just anticipating a potential review comment.

**Comment by brianlheim** [comment]
@muellmusik friendly reminder to label your PRs!

**Comment by muellmusik** [comment]
> A unit test for this raises a sync issue: `Document.new` synchronously adds to the allDocuments collection -- so there's no way here to avoid a hardcoded wait as far as I can see. Just anticipating a potential review comment.

Okay, as I haven't followed that super-closely, what's best practice at the moment? Or does that remain unresolved?

**Comment by jamshark70** [comment]
> Okay, as I haven't followed that super-closely, what's best practice at the moment?

Best practice (only recently discussed actually) is, if possible, to set up a Condition with a function for `test`, where the function answers false while you're still waiting and true when the asynchronous thing is ready to move on, and plug that into `this.wait` (`this` being the UnitTest object). It's not a beautiful implementation but it's better than hard coding a delay (and it becomes a central location to improve later).

If that doesn't apply to a given case, then I don't think there's an established fallback.

But here, e.g. `Condition({ Document.allDocuments.includes(testDoc) })` will be true immediately because Document `prAdd` is called synchronously, without waiting for a signal from the IDE that the document is ready. (I think.) So we'd be stuck with a timed pause, I think.

**Comment by muellmusik** [comment]
> So we'd be stuck with a timed pause, I think.

Okay, so I'm unclear. Should I write a test with a wait (for how long?) or just save it for a better solution? timed pause seems like a recipe for failed builds. Also I'm assuming I can write into ```Platform.defaultTempDir```?

**Comment by jamshark70** [comment]
> Should I write a test with a wait (for how long?) or just save it for a better solution?

I'd say write the test with a wait... how long? Try empirically I guess. If it takes longer than 100 ms for Document.new to complete, then I'd say there are other problems (and probably some other Document tests would fail too).

But you raise a valid question: If unit tests are expected whenever reasonable, and are expected to follow some specific best practices, what is the solution when best practices won't work in a specific case? My opinion is, relax the best practices and have a functioning test. I'm not sure we really have consensus on that, though.

> timed pause seems like a recipe for failed builds.

AFAIK we aren't running the test suite in CI... we're writing these tests so that, say, a developer working on a Document feature can do `TestDocument.run` and have some (higher) degree of confidence that they haven't broken it.

I think CI would be skipping tests that are dependent on the IDE anyway?

> Also I'm assuming I can write into Platform.defaultTempDir?

You should -- I've used that in tests before.

**Comment by jamshark70** [comment]
BTW #4582 adds TestDocument -- after that is merged, you could rebase and add this method to that:

```
	test_save_method_writes_file {
		var doc, file, str, path;
		if(Platform.ideName == "scqt") {
			path = Platform.defaultTempDir +/+ "test_document_save.txt";
			if(File.exists(path)) {
				this.assert(false, "Document save test file path should not exist: '%'".format(path));
			} {
				doc = Document(string: "abcde");
				// note: AFAIK there is no signal coming from IDE to sclang
				// to tell it that the document is ready in the editor.
				// Without a signal, there is no Condition to wait on
				// so we have no choice but to use a timed wait
				0.1.wait;
				doc.save(path);
				0.1.wait;
				file = File(path, "r");
				if(file.isOpen) {
					protect {
						str = file.readAllString;
					} {
						file.close;
						File.delete(path);
					};
					this.assertEquals(str, "abcde", "Document save method should write the correct string to disk");
				} {
					this.assert(false, "Document autoRun test could not open '%' for writing".format(path));
				};
			}
		} {
			// TODO: skip
		};
	}
```

Untested but should be valid.

**Comment by muellmusik** [comment]
thanks James!

**Comment by jamshark70** [comment]
> #4582 adds TestDocument

Just realized, #4582 is a bugfix on 3.10, but this is develop. So TestDocument would have to be introduced here by rebase or cherry-pick. Not hard but worth noting.

**Comment by brianlheim** [comment]
> Just realized, #4582 is a bugfix on 3.10, but this is develop. So TestDocument would have to be introduced here by rebase or cherry-pick. Not hard but worth noting.

that's not how we do it --

> Periodically (roughly every week), the release branch is merged into develop so that develop always has both the latest bug fixes and features.

https://github.com/supercollider/supercollider/wiki/git-workflow-and-guidelines

**Comment by brianlheim** [comment]
travis's push job is failing because this branch diverged before https://github.com/supercollider/supercollider/pull/4691. i'm fine ignoring it, but just noting that if you want to see a green check there you'll have to rebase this branch on the current tip of develop.

File: SCClassLibrary/scide_scqt/ScIDE.sc
Comment (brianlheim) [inline SCClassLibrary/scide_scqt/ScIDE.sc]: `if(docPath.isNil) { MethodError("Document saved requires specified path", this).throw };`

see https://github.com/supercollider/supercollider/wiki/Code-style-guidelines#spaces-in-expressions-and-statements

no newline needed in the error string, at least AFAIK

typically we prefer to notate `if`s like that in the class library, although we don't have a specific rule for that. open to preference i guess

File: editors/sc-ide/core/doc_manager.cpp
Comment (brianlheim) [inline editors/sc-ide/core/doc_manager.cpp]: the code in this file could really do for some refactoring. i also see the following issues --

- no warnings / feedback given for any of the implicit `else` branches of if's
- possible unnecessary string copy
- no feedback given to language on completion

i don't expect you to solve any of that in this PR. after this is merged i wouldn't mind submitting one to work down this technical debt

File: HelpSource/Classes/Document.schelp
Comment (brianlheim) [inline HelpSource/Classes/Document.schelp]: > After a successful save, this Document's path will be updated if needed.

where does that happen?

**Comment by muellmusik** [comment]
> travis's push job is failing because this branch diverged before #4691. i'm fine ignoring it, but just noting that if you want to see a green check there you'll have to rebase this branch on the current tip of develop.

Cheers. I'll do it.

File: HelpSource/Classes/Document.schelp
Comment (muellmusik) [reply HelpSource/Classes/Document.schelp]: On the IDE side in ```DocumentManager::doSaveAs``` which updates the path and then calls ```syncLangDocument```. It would be possible to set the path immediately for synchronous access, but this would make it wrong if the save fails for any reason.

File: SCClassLibrary/scide_scqt/ScIDE.sc
Comment (muellmusik) [reply SCClassLibrary/scide_scqt/ScIDE.sc]: > typically we prefer to notate `if`s like that in the class library, although we don't have a specific rule for that. open to preference i guess

I'm a bit old school I guess. I think in the earlier versions only ```if(conditon, {})``` was allowed in that style. I still prefer it implicitly as it makes clearer that it's a method call, but I accept that's esoteric. ;-) But then I was happy with ```boolean.if(...)``` for this reason though many consider it an abomination, LOL.

**Comment by muellmusik** [comment]
I've not added the test here for reasons discussed above. Would make sense perhaps to update #4582 and then  add it there? I can do this.

**Comment by muellmusik** [comment]
Okay I think this otherwise ready? All tests passing.

**Comment by brianlheim** [comment]
> I've not added the test here for reasons discussed above. Would make sense perhaps to update #4582 and then add it there?

not sure what you mean -- are you suggesting to add the changes from this PR to #4582?

**Comment by muellmusik** [comment]
> EDIT -- very confused what you mean by "for reasons discussed above". why not just add the test here and then resolve merge conflicts when merging 3.10 into develop? (or alternatively wait until that PR is merged, and 3.10 merged to develop, and then rebase this onto the new tip of develop)

The proposed test is a method added to a class being added in #4582. So we can either merge this and then add the test to #4582, or wait for #4582 to be merged as you suggest. The former just seemed cleaner to me in terms of commit order, and I'm happy to add the method there. Not a big deal either way.

**Comment by brianlheim** [comment]
let's not add irrelevant code added to #4582. we can just wait until that is merged

**Comment by jamshark70** [comment]
> we can just wait until that is merged

Agreed -- 1. Merge 4582 into 3.10. 2. Merge 3.10 into develop (per normal practice as discussed above). 3. Rebase this on develop. 4. Add test.

Correct?

**Comment by brianlheim** [comment]
yep!

**Comment by muellmusik** [comment]
Yup, agreed. Sorry, I somehow had the notion that TestDocument was a more extensive class as added in the other PR.

**Comment by brianlheim** [comment]
marking as WIP until #4582 is merged

**Comment by muellmusik** [comment]
@brianlheim just to confirm: #4582 was merged to develop, so no need to merge 3.10 to develop before rebasing this?

**Comment by brianlheim** [comment]
this doesn't need to be rebased at all now!

**Comment by dyfer** [comment]
@muellmusik is this PR waiting for tests to be added?

**Comment by muellmusik** [comment]
Yes, sorry it's likely trivial, but I've not had time.

**Comment by dyfer** [comment]
No problem, just checking what's the status currently.

**Comment by muellmusik** [comment]
> No problem, just checking what's the status currently.

No worries. My work pattern with SC these days is periods of availability for semi-intensive work, followed by being swamped by other stuff. This just went over the boundary! ;-)

I'll try to get to it. If somebody's desparate for this IIRC it's just adding @jamshark70's code above and testing it works.

**Comment by joshpar** [comment]
is this something we still want to pull in? We are trying to clear out PRs for 3.13. Please let us know.
 @muellmusik

**Comment by muellmusik** [comment]
Hi @joshpar Yes ideally. If it’s fine without the test can be merged. Otherwise I might be able to add it tomorrow.

**Comment by muellmusik** [comment]
Okay this has now been rebased on latest develop (resolved a conflict) and added @jamshark70's test method. All seems to run fine. I believe this still shows as needing a review, but Brian did look at it and I made the requested changes. The test method is the only substantive thing that has changed since then. Would be really nice to get this in. :-)

**Comment by muellmusik** [comment]
Anything I need to do to move this along?

File: testsuite/classlibrary/TestDocument.sc
Comment (telephon) [inline testsuite/classlibrary/TestDocument.sc:20]: is this second wait necessary?

File: testsuite/classlibrary/TestDocument.sc
Comment (telephon) [inline testsuite/classlibrary/TestDocument.sc:36]: what is missing here?

File: HelpSource/Classes/Document.schelp
Comment (telephon) [inline HelpSource/Classes/Document.schelp:247]: I like the interface!

File: testsuite/classlibrary/TestDocument.sc
Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc:36]: Yeah, sorry. Again @jamshark70 wrote this but I think it's a placeholder for other editors. Could be removed now.

File: HelpSource/Classes/Document.schelp
Comment (muellmusik) [reply HelpSource/Classes/Document.schelp:247]: Ta!

File: testsuite/classlibrary/TestDocument.sc
Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc:36]: @jamshark70 Any comments on the queries above. If not I will remove the unused false branch.

Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc:36]: Actually, as we're inside a running UnitTest, I'm unsure. In the case that we're outside the IDE, the test just shouldn't run. But there's no clear way to assert that the test is just not applicable. What's the correct practice here? Post a warning and do nothing? Assert true with a warning that the test wasn't run? UnitTest doc doesn't make this clear.

Comment (dyfer) [reply testsuite/classlibrary/TestDocument.sc:36]: > What's the correct practice here? Post a warning and do nothing? Assert true with a warning that the test wasn't run? UnitTest doc doesn't make this clear.

We don't have a specific way to address this ATM. For the test runner used in github actions (QPM) we use a script that specifies which tests to exclude and indeed Document tests are [already excluded](https://github.com/supercollider/supercollider/blob/96eaed0b0c7b72e188f000bdacd237e9418ac3b5/testsuite/scripts/gha_test_run_proto.json#L31-L36). I personally think that we should have a built-in test runner if possible where such configuration would be more flexible and allow for conditional exclusion based on which ide is used etc.

Since these tests are currently excluded in the gha test runner, the question remains what should the test do if it's run outside of QPM and not in IDE. IMO it should _not_ assert false... but aside from that I don't know whether it's better to assert true or not and whether to post something or not. I'd be happy to leave this as is. I'd consider adding a little more to the "todo" comment, like `// TODO: decide what to do if the test is run outside of the scide`. I'd be also fine with adding a warning that the test did not run. Unless others feel strongly otherwise, I wouldn't add assert true since we don't have consistent guidelines atm.

Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc]: Great @dyfer! I've added both your TODO text and a warning. Hopefully we can move this along now?

File: testsuite/classlibrary/TestDocument.sc
Comment (dyfer) [inline testsuite/classlibrary/TestDocument.sc:9]: @muellmusik I think this technically should be an `.scd` document, but that's a really minor thing. Feel free to change if you like, otherwise we can merge as is.

**Review by dyfer — CHANGES_REQUESTED** [review]
I've tested this and it works fine when the provided path is valid and writable. However, `-save` does not provide any feedback if the path is not writable or if the directory doesn't exist - it fails silently.
IMO it should throw an error (in comparison `Document.open("/invalid/path.scd")` throws an error).

**Comment by dyfer** [comment]
@muellmusik I just had one comment regarding the error handling, would you be able to address that?

**Comment by dyfer** [comment]
@muellmusik just wanted to check in, would you be able address the review comments?

**Comment by muellmusik** [comment]
Sorry, @dyfer yes I can do that.

**Comment by muellmusik** [comment]
> I've tested this and it works fine when the provided path is valid and writable. However, `-save` does not provide any feedback if the path is not writable or if the directory doesn't exist - it fails silently.
> IMO it should throw an error (in comparison `Document.open("/invalid/path.scd")` throws an error).

Okay it throws an error now if the directory doesn't exist. I think the Qt side should handle any other errors.

File: SCClassLibrary/scide_scqt/ScIDE.sc
Comment (telephon) [inline SCClassLibrary/scide_scqt/ScIDE.sc:640]: would we want to throw that error? Not sure myself, just wondering.

**Comment by dyfer** [comment]
Thanks @muellmusik !
I think the error should really be thrown on the C++ side. I'd consider a situation to manually catch a particular scenario in sclang to be acceptable only as a stop gap solution, identifying the underlying issue in C++.
Do we know why there's no error thrown in C++ when the saving fails?

File: SCClassLibrary/scide_scqt/ScIDE.sc
Comment (dyfer) [reply SCClassLibrary/scide_scqt/ScIDE.sc:640]: As I wrote in the comment, I think the errors should actually be thrown on the c++ side, as opposed to trying to manually check for a limited set of conditions.

Comment (muellmusik) [reply SCClassLibrary/scide_scqt/ScIDE.sc:640]: As to whether it should be thrown or not I can imagine people may want to catch it.

**Comment by muellmusik** [comment]
While I see your general point @dyfer, perhaps you could say a little more about how you think that would work? It's been a while since I've been immersed in this part of the code base, so it's completely possible I've missed something, but AFAICS on a quick look, there is no current way to throw an error back across the IPC bridge. I think any such approach would have to either block (bad I think) or be asynchronous (weird). 

As it is I modelled it on the error handling in open which is a lang check as well. There's actually a lot of this sort of thing because of the need to sync lang and IDE doc states anyway. A lang check, while inelegant, is at least reliable and has the virtue of simplicity and standard behaviour.

Happy to consider another approach though if someone can explain how it should work!

**Comment by muellmusik** [comment]
> in particular, the current solution won't catch an issue I encountered, where the specified directory exists, but is not writable

I'm not sure why that doesn't at least post something. It uses the same catch block with a qWarning as open and other IDE handler methods.

```
    } catch (std::exception const& e) {
         qWarning() << "DocumentManager::" << __FUNCTION__ << ": could not handle request:" << e.what();
         return;
     }
 }
```

If that's not working it seems like a more general problem?

**Comment by dyfer** [comment]
>> If that's not working it seems like a more general problem?

Yes, that seems to be an issue - at least it should post something.

I see the larger point about IPC communication, thank you for clarifying. My bad, I thought we were calling a primitive directly here, I'm sorry for the noise. But it's good that there's clarity about things that don't work.

In the case of this PR, I would suggest to indicate a `NOTE::` in the help that the save method might fail silently (under the `-save` entry).

Since there's no way to "properly" catch an issue here, I think we could remove the sclang check since it won't catch some access issues anyway... However, I'm okay with it staying. I'd just add a note that this ideally should be communicated from the IDE code.

I'd file a separate issue about lack of errors for `save` (and any related methods, if that's the case) - I'd be happy to do that.

Finally, some programs automatically create necessary folders if saving is requested in a specific location that doesn't exist. Would that be a desirable behavior? This could be something for another Issue or PR.

**Comment by muellmusik** [comment]
@dyfer shall I go ahead with the directory creation?

File: SCClassLibrary/scide_scqt/ScIDE.sc
Comment (dyfer) [reply SCClassLibrary/scide_scqt/ScIDE.sc:640]: I feel this should throw an error, more below.

File: testsuite/classlibrary/TestDocument.sc
Comment (dyfer) [reply testsuite/classlibrary/TestDocument.sc:9]: @muellmusik do you think this could be changed to `.scd` or do you want to leave it as is?

**Comment by dyfer** [comment]
> I'm not sure why that doesn't at least post something. It uses the same catch block with a qWarning as open and other IDE handler methods.

BTW an error is indeed posted (not exactly from the place you linked, but it is there nonetheless) if the writing fails, but it is posted in the stdout of the IDE process, i.e. you can see it only if you run the IDE from the terminal; i.e. it's not posted in the sclang post window. I don't think we can use it for anything, but I thought I'd point out what's happening here.

File: HelpSource/Classes/Document.schelp
Comment (dyfer) [inline HelpSource/Classes/Document.schelp:242]: An open question: do you think it would be worthwhile to also link to this method right below the `Document.open` entry? It's really far down from the `open` class method and the top of the file, and one could argue they are related (even though the first one is class method, the other an instance method). OTOH it can be found if you search for `save` or trigger help for the method name in the first place... so maybe that's not needed? This came to me when I was scrolling through the help file trying to find the method manually.

Comment (muellmusik) [reply HelpSource/Classes/Document.schelp:242]: Don't mind. Can just move it if that makes more sense?

**Comment by muellmusik** [comment]
> I needed to check what's the behavior of other file methods (`File` and `Object -writeArchive`). In both cases the primitive fails (Error thrown) in case when directory doesn't exist _and_ if the directory exists but it's not writeable.

I believe these work as their lang checks, not IDE related.


> * check path for nil, and then use a system command to check whether the directory is writable (that would also fail if the directory doesn't exist)

That is heavyweight. If we go that route, would be nice if there was a cross-platform way of doing it, so we don't get the multi-platform technical debt. Maybe something is in std::filesystem or boost? Then we could implement a lang primitive for that query.

**Comment by muellmusik** [comment]
I think maybe if this returns file_type::unknown? https://en.cppreference.com/w/cpp/filesystem/status

File: HelpSource/Classes/Document.schelp
Comment (dyfer) [reply HelpSource/Classes/Document.schelp:242]: It's tricky! `save` is an instance method, while `open` is a class method... So while I'd like to see them together, this would be against the existing convention of separating class and instance methods. That's why I think a link would be better.

**Comment by dyfer** [comment]
> I believe these work as their lang checks, not IDE related.

Yes, of course. 
My point is that, as a user, I'd expect writing files to behave the same, whether that's writing a text file with some strings or saving the document.

> That is heavyweight. If we go that route, would be nice if there was a cross-platform way of doing it, so we don't get the multi-platform technical debt.

Yes, you're probably right. A new primitive would be better than some clumsy terminal calls.

In that case I suggest we leave this in the current state (first nil check, then directory existence check), but I still feel the directory check should _throw_ an error. You could also add a comment that ideally this should be replaced by a "is directory writable" check in the future.

**Comment by muellmusik** [comment]
> You could also add a comment that ideally this should be replaced by a "is directory writable" check in the future.

Okay added. Is this ready to merge?

**Comment by dyfer** [comment]
> Okay added. Is this ready to merge?

Sorry, but no. The directory check still does not throw an error, which was my request from [earlier](https://github.com/supercollider/supercollider/pull/4696#discussion_r1152263932).

I also suggested to add a link in the helpfile, and to rename the test file to *.scd. These two are just suggestions, but it's not clear to me whether you don't want to add them or if they have fallen off the radar.

File: testsuite/classlibrary/TestDocument.sc
Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc:9]: What do you mean? You cannot declare classes in an scd file, so?

File: HelpSource/Classes/Document.schelp
Comment (muellmusik) [reply HelpSource/Classes/Document.schelp:243]: Done. Also did a link from save back to *open.

**Comment by muellmusik** [comment]
I think those are all addressed now. Sorry about the error thing. Forgot about String:error.

File: testsuite/classlibrary/TestDocument.sc
Comment (dyfer) [reply testsuite/classlibrary/TestDocument.sc:20]: BTW these waits are okay, i.e. there's no way to avoid them, AFAIU (due to async communication sclang <-> IDE).
Also, these tests won't be run in the automated test suite in GitHub Actions for the time being, since we don't run them in the IDE (just in sclang alone) and thus `Document` tests are disabled.

File: testsuite/classlibrary/TestDocument.sc
Comment (muellmusik) [reply testsuite/classlibrary/TestDocument.sc]: Ah okay, sorry. Not sure why I did that tbh! Fixed.

**Comment by muellmusik** [comment]
This still shows as requesting a change, though I think I've resolved everything? IAC @dyfer I think this can be merged.

**Pull request merged**
